### PR TITLE
ISPN-12156 Security Realm identity caching

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.redhat.com/browse/ISPN-12156

### CODE REVIEW

File: server/runtime/src/main/java/org/infinispan/server/configuration/security/RealmConfiguration.java [hunk 1]
```diff
@@ -15,9 +15,11 @@
  */
 public class RealmConfiguration implements ConfigurationInfo {
    static final AttributeDefinition<String> NAME = AttributeDefinition.builder("name", null, String.class).build();
+   static final AttributeDefinition<Integer> CACHE_MAX_SIZE = AttributeDefinition.builder("cacheMaxSize", 16).build();
```
Comment (danberindei) [inline server/runtime/src/main/java/org/infinispan/server/configuration/security/RealmConfiguration.java:18]: Isn't 16 kind of small?

File: server/runtime/src/main/java/org/infinispan/server/configuration/security/RealmConfigurationBuilder.java [hunk 2]
```diff
@@ -157,4 +179,35 @@ SSLContext getSSLContext() {
...
+   public void addRealm(String realmName, SecurityRealm realm) {
+      addRealm(realmName, realm, null);
+   }
+
+   public void addRealm(String realmName, SecurityRealm realm, Consumer<SecurityDomain.RealmBuilder> realmBuilderConsumer) {
+      SecurityDomain.RealmBuilder realmBuilder = domainBuilder.addRealm(realmName, cacheable(realm));
+      if (realmBuilderConsumer != null) {
+         realmBuilderConsumer.accept(realmBuilder);
+      }
+      realmBuilder.build();
```
Comment (danberindei) [inline server/runtime/src/main/java/org/infinispan/server/configuration/security/RealmConfigurationBuilder.java:208]: I know it's not our API but I feel like it merits a comment, seeing the result of `build()` unused is puzzling.

File: server/runtime/src/main/java/org/infinispan/server/configuration/security/RealmConfiguration.java (hunk 1 above)
Comment (gustavonalle) [reply server/runtime/src/main/java/org/infinispan/server/configuration/security/RealmConfiguration.java:18]: It seems so. If I understand it correctly, this cache is mapped 1:1 to user credentials in the realm. If the entries are small, we could maybe up this number?

File: server/tests/src/test/java/org/infinispan/server/security/authentication/RestAuthentication.java [hunk 3]
```diff
@@ -78,7 +78,7 @@ public void testStaticResourcesAnonymously() {
 
    @Test
    public void testRestReadWrite() {
-      RestClientConfigurationBuilder builder = new RestClientConfigurationBuilder();
+      RestClientConfigurationBuilder builder = new RestClientConfigurationBuilder().socketTimeout(100_000).connectionTimeout(100_000);
```
Comment (gustavonalle) [inline server/tests/src/test/java/org/infinispan/server/security/authentication/RestAuthentication.java:81]: What is the reason for this change?

**Comment by gustavonalle** [comment]
Ping @tristantarrant could you address the review issues?

File: server/runtime/src/main/java/org/infinispan/server/configuration/security/RealmConfiguration.java (hunk 1 above)
Comment (tristantarrant) [reply server/runtime/src/main/java/org/infinispan/server/configuration/security/RealmConfiguration.java:18]: To be honest, I was thinking that many systems will only have a handful of application users.

**Comment by tristantarrant** [comment]
I believe I have addressed all issues

**Comment by gustavonalle** [comment]
Checkstyle error

**Comment by tristantarrant** [comment]
Fixed

**Comment by tristantarrant** [comment]
This is failing because of a bug in Elytron: https://issues.redhat.com/browse/ELY-2031

**Comment by tristantarrant** [comment]
I've created a PR for Elytron https://github.com/wildfly-security/wildfly-elytron/pull/1456 that fixes the issue.

File: server/runtime/src/main/java/org/infinispan/server/security/realm/CachingSecurityRealm.java [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+
+         @Override
+         public boolean exists() throws RealmUnavailableException {
+            return true; // non-existing identities will not be wrapped
+         }
+
+         @Override
+         public AuthorizationIdentity getAuthorizationIdentity() throws RealmUnavailableException {
+            if (authorizationIdentity == null) {
+               log.tracef("getAuthorizationIdentity Caching AuthorizationIdentity for principal='%s'", principal);
```
Comment (danberindei) [inline server/runtime/src/main/java/org/infinispan/server/security/realm/CachingSecurityRealm.java:238]: Missing `isTraceEnabled()` (in getAttributes() as well)

**Review by danberindei — CHANGES_REQUESTED** [review]

File: server/rest/src/main/java/org/infinispan/rest/RestRequestHandler.java [hunk 5]
```diff
@@ -124,6 +131,19 @@ public void channelRead0(ChannelHandlerContext ctx, FullHttpRequest request) thr
       });
    }
 
+   private boolean matchAuthorizationHeader(String authz, String cached) {
```
Comment (wburns) [inline server/rest/src/main/java/org/infinispan/rest/RestRequestHandler.java:134]: Seems this is unused?

File: server/runtime/src/main/resources/schema/infinispan-server-12.0.xsd [hunk 6]
```diff
@@ -820,6 +820,20 @@
          <xs:element type="tns:security-realm-truststore" name="truststore-realm" minOccurs="0"/>
       </xs:sequence>
       <xs:attribute name="name" type="xs:string" use="required"/>
+      <xs:attribute name="cache-max-size" type="xs:int" default="16">
```
Comment (wburns) [inline server/runtime/src/main/resources/schema/infinispan-server-12.0.xsd:823]: Looks like the code says 256?

**Comment by tristantarrant** [comment]
Elytron have finally merged my changes and released a new version. I've addressed all comments.

**Comment by tristantarrant** [comment]
Rebased

File: server/runtime/src/main/java/org/infinispan/server/security/realm/CachingSecurityRealm.java [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+    *
+    * @param realm the {@link SecurityRealm} whose {@link RealmIdentity} should be cached.
+    * @param cache the {@link RealmIdentityCache} instance
+    */
+   public CachingSecurityRealm(CacheableSecurityRealm realm, RealmIdentityCache cache) {
+      this.realm = checkNotNullParam("realm", realm);
+      this.cache = checkNotNullParam("cache", cache);
+
+      if (realm instanceof CacheableSecurityRealm) {
+         CacheableSecurityRealm cacheable = (CacheableSecurityRealm) realm;
```
Comment (danberindei) [inline server/runtime/src/main/java/org/infinispan/server/security/realm/CachingSecurityRealm.java:48]: Cast is not needed

File: server/runtime/src/main/java/org/infinispan/server/security/realm/CachingSecurityRealm.java [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+         return realmIdentity;
+      }
+
+      RealmIdentity cachedIdentity = new RealmIdentity() {
+         final RealmIdentity identity = realmIdentity;
+
+         AuthorizationIdentity authorizationIdentity = null;
+         Attributes attributes = null;
+         IdentityCredentials credentials = IdentityCredentials.NONE;
+         int verifiedEvidenceHash;
```
Comment (danberindei) [inline server/runtime/src/main/java/org/infinispan/server/security/realm/CachingSecurityRealm.java:81]: So it's not enough to cache the `RealmIdentity`, we also have to cache the `AuthorizationIdentity` and `IdentityCredentials`?

**Review by danberindei — CHANGES_REQUESTED** [review]

**Comment by danberindei** [comment]
Merged, thanks Tristan!
